### PR TITLE
bump cpp-httplib version to v0.7.6 latest

### PR DESCRIPTION
**Pull request by breznak**
hoping it might resolve issue in #783

### CODE REVIEW

**Comment by breznak** [comment]
Locally, I'm getting sign-compare errors in the lib, which we don't want to fix directly. 

> [ 57%] Building CXX object src/test/CMakeFiles/unit_tests.dir/unit/regions/SPRegionTest.cpp.o
In file included from /mnt/store/devel/HTM/htm-community/nupic.cpp/src/test/unit/engine/RESTapiTest.cpp:34:
/mnt/store/devel/HTM/htm-community/nupic.cpp/build/ThirdParty/cpp-httplib/cpp-httplib-src/httplib.h: In function ‘bool httplib::detail::contains_header(const string&, const string&)’:
/mnt/store/devel/HTM/htm-community/nupic.cpp/build/ThirdParty/cpp-httplib/cpp-httplib-src/httplib.h:2840:23: error: comparison of integer expressions of different signedness: ‘int’ and ‘std::__cxx11::basic_string<char>::size_type’ {aka ‘long unsigned int’} [-Werror=sign-compare]
2840 |     for (int i = 0; i < name.length(); ++i) {

Is the `-isystem` include broken again?

**Review by dkeeney — DISMISSED** [review]
Yes, this looks good.

**Comment by breznak** [comment]
https://stackoverflow.com/questions/3371127/use-isystem-instead-of-i-with-cmake 

but I'm not sure where we should apply it (ideally all externals use `SYSTEM` include)

**Comment by Zbysekz** [comment]
Update on 0.7.6 didn't solve the issue on my PC, test keeps failing.

Note: I have to correct the signess in the for loop in the library header & comment client->set_timeout_sec(30); everywhere on our side since they changed this function

**Comment by breznak** [comment]
> Note: I have to correct the signess in the for loop in the library header & comment client->set_timeout_sec(30); everywhere on our side since they changed this function

thanks for reporting. sad it didn't resolve the issue. Do we want to go through updating our codebase to this anyway?

**Comment by dkeeney** [comment]
Oh, I think I see it...
This error for the sign compare occurs where we compile the client and the server for the RESTapi.  It is not part of the htm.core library.  On line 450 and 464 of  src/CMakeLists.txt,  I did not specify SYSTEM for EXTERNAL_INCLUDES.

Also line 432 for the MNIST and other places.

We could catch all of them in one place by including SYSTEM at the beginning of the list at the point where the EXTERNAL_INCLUDES list is built which is line 93 of extern/bootstrap.cmake.  If we do that, the SYSTEM at line 249 of src/CMakeLists.txt becomes redundant so that could be removed.

I have not tried this but it should work.

File: bindings/py/cpp_src/CMakeLists.txt [hunk 1]
```diff
@@ -132,11 +132,11 @@ target_link_libraries(${algorithms_shared_lib} PRIVATE
 target_compile_options(${algorithms_shared_lib} PUBLIC ${INTERNAL_CXX_FLAGS})
 target_compile_definitions(${algorithms_shared_lib} PRIVATE ${COMMON_COMPILER_DEFINITIONS})
 target_include_directories(${algorithms_shared_lib} PRIVATE 
-        ${PYTHON_INCLUDE_DIRS}
-        ${pybind11_INCLUDE_DIRS}
-        ${CORE_LIB_INCLUDES} 
-        ${EXTERNAL_INCLUDES}
+        ${CORE_LIB_INCLUDES}
         ${PROJECT_SOURCE_DIR}
+	SYSTEM ${PYTHON_INCLUDE_DIRS}
```
Comment (breznak) [inline bindings/py/cpp_src/CMakeLists.txt:137]: Thanks @dkeeney for the advice!, 

> We could catch all of them in one place by including SYSTEM at the beginning of the list at the point where the EXTERNAL_INCLUDES list is built which is line 93 of extern/bootstrap.cmake.  

I tried this, by adding `"SYSTEM"` to the `set(EXTERNAL_INCLUDES ...) ` but that didn't work as intended. 

> On line 450 and 464 of src/CMakeLists.txt, I did not specify SYSTEM for EXTERNAL_INCLUDES. 

..so I fixed all ocurences of EXTERNAL_INCLUDES in `target_include_dirs()` with SYSTEM, that works. 
But not for `httplib`, unfortunately. Is it included in the EXTERNAL_INCLUDES, or does it use a different means of including?

File: src/examples/rest/client.cpp [hunk 2]
```diff
@@ -37,10 +37,17 @@
...
 
-#include <httplib.h>
 #include <iostream>
 #include <cstring>
 
+// save diagnostic state
+#pragma GCC diagnostic push
+// turn off the specific warning. Can also use "-Wall"
+#pragma GCC diagnostic ignored "-Wsign-compare"
+#include <httplib.h>
```
Comment (breznak) [inline src/examples/rest/client.cpp:47]: the `SYSTEM` include did not fix the warnings generated in httplib (I wonder why?), so I had to hack with `#pragma` here and in `server_core.hpp`.

File: src/examples/rest/client.cpp [hunk 3]
```diff
@@ -78,7 +85,7 @@ int main(int argc, char **argv) {
 
   VERBOSE << "Connecting to server: " + serverHost + " port: " << port << std::endl;
   httplib::Client client(serverHost.c_str(), port);
-  client.set_timeout_sec(30);  // The time it waits for a network connection.
+  //client.set_timeout_sec(30);  // The time it waits for a network connection.
```
Comment (breznak) [inline src/examples/rest/client.cpp:88]: @Zbysekz did you find out if there's a replacement for the `httplib.set_timeout_sec()`, or if we just stop using that function?

**Review by breznak — COMMENTED** [review]
It now seems to be working with the new lib. Please see comments bellow and possibly approve.

File: src/examples/rest/client.cpp (hunk 3 above)
Comment (Zbysekz) [reply src/examples/rest/client.cpp:88]: Asked question [https://github.com/yhirose/cpp-httplib/issues/642](https://github.com/yhirose/cpp-httplib/issues/642)

File: src/examples/rest/client.cpp (hunk 2 above)
Comment (dkeeney) [reply src/examples/rest/client.cpp:47]: > the SYSTEM include did not fix the warnings

I don't know of a way to get around the warnings other than with the #pragma.

**Review by dkeeney — APPROVED** [review]
Looks good.

File: src/examples/rest/client.cpp (hunk 3 above)
Comment (Zbysekz) [reply src/examples/rest/client.cpp:88]: We have an answer: `set_connection_timeout()` is the substitute.
@breznak this PR is closed maybe starting another? I am not sure about consequences, or what timeout is by default

Comment (breznak) [reply src/examples/rest/client.cpp:88]: nice! sorry for merging early. Yes, we can start a new PR where you can revert commit `606cb75d00cb376` and then change it to the new signature. 

> I am not sure about consequences, or what timeout is by default 

nor do i. our tests passed even without it. maybe in the server it makes sense, in the unit tests 30sec is imho too much.

**Comment by dkeeney** [comment]
>  maybe in the server it makes sense, in the unit tests 30sec is imho too much.

The default web client timeout is 2 min (same as almost all web browsers).  The call to set_timeout_sec( ) was an attempt to shorten the time the client takes to report that it cannot find the server.  If the connection happens right away the timeout setting has no effect.  I think we can safely remove the call to set the timeout.

**Comment by breznak** [comment]
> I think we can safely remove the call to set the timeout. 

thank you for explanation David. So then I think no further action is needed and this can stay as is.

**Pull request merged**